blogs/ulysses-offload/README.md

```diff
@@ -0,0 +1,263 @@
+and subsequent sections provide additional details on the design and 
+usability of FPDT, followed by experimental results.
+
+## Key Innovations
+
+### 1. Fully Pipelined Distributed Transformer (FPDT)
+
+The core innovation of our work is the Fully Pipelined Distributed
+Transformer (FPDT). This approach leverages a pipelined sequence
+chunking, which allows for the training of LLMs with sequence lengths up
+to 2 million tokens on just 4 A100-40GB GPUs. By breaking down the
+sequence into manageable chunks and processing them in a pipelined
+manner, FPDT significantly reduces the memory footprint while
+maintaining high computational efficiency. This method ensures that the
+GPUs are utilized effectively, even when dealing with extremely long
+sequences.
+
+### 2. Memory Optimization
+
+One of the critical aspects of our approach is the comprehensive
+analysis and optimization of the memory footprint during LLM training.
+We target the reduction of redundant intermediate buffers in both the
+forward and backward passes of the training process. By optimizing the
+use of GPU and host CPU memory, we can train larger models with longer
+sequences without running into GPU memory limitations. This optimization
+is crucial for enabling the training of ultra-long context models on a
+limited number of GPUs. It is worth noting that FPDT memory optimization
+is orthogonal and complementary to model- parameter-focused memory
+optimization techniques used by DeepSpeed ZeRO and PyTorch FSDP.
+
+### 3. Compatibility and Flexibility
+
+FPDT is designed to be agnostic to existing training techniques and
+works efficiently across different LLM models, including popular
+architecture like GPT and Llama. This flexibility ensures that our
+approach can be easily integrated into various training workflows.
+Additionally, FPDT is compatible with advanced memory optimization
+techniques such as DeepSpeed ZeRO and PyTorch FSDP, further enhancing
+its usability and performance.
+
+## Core Design of Fully Pipelined Distributed Transformer
+
+Figure 2 illustrates the core structure of FPDT. FPDT leverages multiple
+memory hierarchies in modern GPU clusters, thus boosting hardware
+efficiency and cost-effectiveness while achieving very high model FLOP
+utilization (MFU). The design of FPDT centers around pipelining,
+scheduling, and memory management. These well-known optimization
+techniques are essential for scaling LLM context length to a million
+scale with a few GPUs and will be discussed in the subsequent
+subsections.
+
+<img src="./media/image2.png" style="width:6.5in;height:2.68634in"
+alt="A screenshot of a computer Description automatically generated" />
+
+Figure 2: FPDT core design
+
+### 
+
+### Pipelining and Scheduling
+
+FPDT employs a pipelined sequence chunking design to manage the memory
+and computational load efficiently. In traditional Transformer model,
+input QKV tensor can be denoted *\[B, S, H, D\]*, where *B* is batch
```

Comment:
QKV tensor => q, k, v (separate) tensors. Ulysses assumes input sequence/hidden state projected so q,k,v!

Reply:
The confusion is the overload of `input` since the general conversation is about sequences, which are also input. How about using `projected` instead of `input`? 

Also, does [B, S, H, D] denote a fused QKV tensor, or separate Q, K, V tensors`?

Reply:
Separate